### PR TITLE
Add pvr.octonet plugin

### PR DESCRIPTION
**Pull request by julianscheel**
Add new pvr plugin for DigitalDevices Octonet Sat>IP servers.

Signed-off-by: Julian Scheel <julian@jusst.de>

### CODE REVIEW

**Comment by Jalle19** [comment]
@ksooo any chance you could test this addon since you have the hardware?

**Comment by ksooo** [comment]
> any chance you could test this addon since you have the hardware?

 sure. will do the next days.

File: pvr.octonet/pvr.octonet.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+pvr.octonet https://github.com/DigitalDevices/pvr.octonet master
```
Comment (ksooo) [inline pvr.octonet/pvr.octonet.txt]: Please, point to a specific release tag here, not to master (or another branch).

**Review by ksooo — CHANGES_REQUESTED** [review]

File: pvr.octonet/pvr.octonet.txt (hunk 1 above)
Comment (julianscheel) [reply pvr.octonet/pvr.octonet.txt]: Sure, I'll change that.

**Comment by julianscheel** [comment]
@ksoo It's pointing to a tag now.
In case you try to test, some notes:
You need to have made a channel scan via the webinterface of the Octonet. We intend to add channel scan control within the plugin but did not do so yet. To see EPG data you also need to manually trigger the EPG scan in the Octonet's webinterface. This is about to change with an upcoming firmware release from DigitalDevices, which will include an automatic background scan for EPG data.

**Comment by ksooo** [comment]
@julianscheel thanks for the howto.

Regarding the tag: https://github.com/DigitalDevices/pvr.octonet/blob/0.1/pvr.octonet/addon.xml#L4 says version of the addon is 0.3 Shouldn't this better be aligned with the tag? I know technically this is not a problem, but still...

**Comment by MartijnKaijser** [comment]
Version in addon.xml should follow x.y.z style so it should become  0.3.0

Edit:
imo the tag should also reflect the Addon version :)

**Comment by ksooo** [comment]
@julianscheel don't laugh at me, but how did you manage to build this on Linux? Having my difficulties with the kodi binary addon build system, tbh. It seems to be impossible to build the new addon without having it added to repo-binary-addons first - so we have a chicken-egg problem I fear... @fetzerch

**Comment by wsnipex** [comment]
@ksooo make sure jsoncpp and kodi headers are installed in a location cmake will find, e.g. by using our ppa packages kodi-addon-dev, kodi-pvr-dev, libjsoncpp-dev
git clone https://github.com/DigitalDevices/pvr.octonet.git
cd pvr.octonet
mkdir build ; cd build
cmake ..
make -j4
sudo make install

**Comment by ksooo** [comment]
@wsnipex have you actually checked that this works for add-ons not yet added to repo-binary-addons? For add-ons already present there building is no problem (for me).

**Comment by ksooo** [comment]
@wsnipex ^^^

**Comment by wsnipex** [comment]
ofc it works.
Building like this(we call it standalone) doesn't care about the addon repos at all.

**Comment by ksooo** [comment]
sniff. ofc my fault, because I'm using nightly and 16.10?

<pre>
kai@kso-desktop:~/src/github/DigitalDevices/pvr.octonet/build$ sudo apt-get install kodi-addon-dev kodi-pvr-dev libjsoncpp-dev
[sudo] password for kai: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  kodi-audio-dev kodi-screensaver-dev kodi-visualization-dev
The following NEW packages will be installed:
  kodi-addon-dev kodi-audio-dev kodi-pvr-dev kodi-screensaver-dev kodi-visualization-dev libjsoncpp-dev
0 upgraded, 6 newly installed, 0 to remove and 9 not upgraded.
Need to get 375 kB of archives.
After this operation, 1.773 kB of additional disk space will be used.
Do you want to continue? [Y/n] y
Ign:1 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main amd64 kodi-addon-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
Ign:2 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main amd64 kodi-audio-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
Ign:3 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main amd64 kodi-pvr-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
Ign:4 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main amd64 kodi-screensaver-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
Ign:5 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main amd64 kodi-visualization-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
Err:1 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main i386 kodi-addon-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
  404  Not Found
Err:2 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main i386 kodi-audio-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
  404  Not Found
Err:3 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main i386 kodi-pvr-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
  404  Not Found
Err:4 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main i386 kodi-screensaver-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
  404  Not Found
Get:6 http://de.archive.ubuntu.com/ubuntu yakkety/main amd64 libjsoncpp-dev amd64 1.7.2-1 [83,5 kB]
Err:5 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu yakkety/main i386 kodi-visualization-dev all 2:17.0~git20161105.0200-3f627a5-0yakkety
  404  Not Found
Fetched 83,5 kB in 0s (250 kB/s)                              
E: Failed to fetch http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu/pool/main/k/kodi/kodi-addon-dev_17.0~git20161105.0200-3f627a5-0yakkety_all.deb  404  Not Found
E: Failed to fetch http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu/pool/main/k/kodi/kodi-audio-dev_17.0~git20161105.0200-3f627a5-0yakkety_all.deb  404  Not Found
E: Failed to fetch http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu/pool/main/k/kodi/kodi-pvr-dev_17.0~git20161105.0200-3f627a5-0yakkety_all.deb  404  Not Found
E: Failed to fetch http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu/pool/main/k/kodi/kodi-screensaver-dev_17.0~git20161105.0200-3f627a5-0yakkety_all.deb  404  Not Found
E: Failed to fetch http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu/pool/main/k/kodi/kodi-visualization-dev_17.0~git20161105.0200-3f627a5-0yakkety_all.deb  404  Not Found
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
</pre>

**Comment by wsnipex** [comment]
sudo apt update
your package cache is old, current kodi package for 16.10 is https://launchpad.net/~team-xbmc/+archive/ubuntu/xbmc-nightly/+packages?field.name_filter=kodi&field.status_filter=published&field.series_filter=yakkety

**Comment by ksooo** [comment]
thanks @wsnipex for all your help, now kodi-addon-dev, kodi-pvr-dev and libjsoncpp-dev are installed and I'm stuck at "cmake .."

<pre>
kai@kso-desktop:~/src/github/DigitalDevices/pvr.octonet/build$ cmake ..
CMake Error at CMakeLists.txt:9 (find_package):
  By not providing "Findkodiplatform.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "kodiplatform", but CMake did not find one.
  Could not find a package configuration file provided by "kodiplatform" with
  any of the following names:
    kodiplatformConfig.cmake
    kodiplatform-config.cmake
  Add the installation prefix of "kodiplatform" to CMAKE_PREFIX_PATH or set
  "kodiplatform_DIR" to a directory containing one of the above files.  If
  "kodiplatform" provides a separate development package or SDK, be sure it
  has been installed.
-- Configuring incomplete, errors occurred!
See also "/home/kai/src/github/DigitalDevices/pvr.octonet/build/CMakeFiles/CMakeOutput.log".
</pre>

**Comment by ksooo** [comment]
I managed to compile it now, with the help of a hack @fetzerch provided. ;-) I can confirm that the addon works quite good. Some glitches here and there, but well ... version 0.3

@julianscheel if you correct the addon version number (like 0.3.0), provide a matching release tag and update this PR accordingly we are ready to merge.

Some screenshots:
screenshot004
screenshot003
screenshot002
screenshot001
screenshot000

**Comment by ksooo** [comment]
@julianscheel look at the screen shots. maybe it's a good idea to fix the name of the company? ;-) DigitalDevice vs DigitalDevices. And I'm not sure whether the product is officially called "Octonet". You probably should ask Bernhard Hieke about this. I thinks it's "Octopus NET"...

EDIT: May I suggest to call the addon "Octopus NET client", maybe "DigitalDevices Octopus NET client"? In no case trailing "PVR" in the titel, please.

EDIT2: Please remove this annoying popup on addon startup saying like "603 channels loaded".

**Comment by wsnipex** [comment]
@ksooo for the record: you were also missing libkodiplatform-dev

**Comment by julianscheel** [comment]
@ksooo Thanks for testing and spotting the typo. I'll fix that, the versioning and will add reporting of signal stats, then we're ready to merge :)
Do distributions (and windows installer builds) automatically adapt plugins added to repo-binary-addons? Or do we manually need to trigger inclusion into those?

**Comment by ksooo** [comment]
> @ksooo for the record: you were also missing libkodiplatform-dev

 @wsinpex: okay. is there some documentation for this, maybe in the wiki? i would be happy to bookmark that page.

**Comment by ksooo** [comment]
> Do distributions (and windows installer builds) automatically adapt plugins added to repo-binary-addons? Or do we manually need to trigger inclusion into those?

 The addon version pointed to by this repo will be automatically pulled into official kodi builds and put into kodi installation packages. The addon maintainer (you, in the case of octonet!) is responsible for keeping an eye on addon's health, in the sense of regularly checking whether it still builds, doing mandatory addon api changes, answering questions in the forum, fixing bugs, etc. If the addon does not build it will "simply" not be packaged (and people start complaining soon in the forum). If team kodi decides that the quality of the addon is insufficient or that the maintainer does not do it's job (anymore), the addon can be removed by team kodi from this repo at any time.

 If you want a new addon version included in the official kodi packages, please file a PR against this repo which updates the respective version tag in pvr.octonet.txt

**Comment by julianscheel** [comment]
@ksooo Thanks for the details. Keeping an eye on the addons health is self-evident :)

**Comment by ksooo** [comment]
> Keeping an eye on the addons health is self-evident :)

 cool. good to know that we're on the same page.

**Comment by wsnipex** [comment]
Note about packages on linux: We only support ubuntu packages in our launchpad PPAs, but for that to work, debian packaging stuff needs to be added to your addon

**Comment by julianscheel** [comment]
@wsnipex Alright, I'll add debian build files to the repository

**Comment by julianscheel** [comment]
@wsnipex I added the debian package stuff to my tree now, but there's one thing I'm wondering about in the other pvr plugin's install files. They all have the following rule for installing the so files:
`usr/lib/*/addons/pvr.iptvsimple/pvr.iptvsimple.so* usr/lib/kodi/addons/pvr.iptvsimple`

Doing a cmake build does install the pvr.octonet.so under the path `usr/lib/x86_64-linux-gnu/kodi/addons/pvr.octonet/`, which is one level deeper than what the pattern from the install file matches. As a wildcard for "kodi" wouldn't make sense I assumed that it is for the architecture and kodi is simply missing in the path, so in my install file I've changed it to:
`usr/lib/*/kodi/addons/pvr.octonet/pvr.octonet.so* usr/lib/kodi/addons/pvr.octonet`
That way I can build a deb package for Ubuntu 16.04 with PackPack.

Do you see any issue with that? If not I'll push it out that way. Who could be asked if it does actually work in the environment that the ubuntu ppa is built in?

**Comment by wsnipex** [comment]
yes, most other addons need to be adapted to the new multi-arch install dir.
the install files should simply contain:

> usr/lib
> usr/share

**Comment by julianscheel** [comment]
I've updated the pull request, pointing to version tag 0.3.2 now. Besides fixing what you mentioned I've added debian package files and Dennis added reporting of signal statistics.

**Comment by Jalle19** [comment]
Should we keep using the `DigitalDevices` GitHub organization on should it be transferred to `kodi-pvr` like our other official PVR addons?

**Comment by julianscheel** [comment]
@Jalle19 I'd be happy with both of it. Maybe moving it to kodi-pvr/ is nicer, because people tend to search for plugins in there.

**Comment by MartijnKaijser** [comment]
@Jalle19 i talked about this with @ksooo and the idea is to just let "companies"  do their own repo and they request an updated with the tag. Doing it this way we don't have to bother with it and save the silly creating PRs of changing code

**Comment by wsnipex** [comment]
debian packaging looks good

**Review by ksooo — APPROVED** [review]

**Comment by MartijnKaijser** [comment]
Lets get It in

**Comment by julianscheel** [comment]
Thanks for merging :)

**Comment by MartijnKaijser** [comment]
Do you want a section on our forum between the other Addons where you can do your support and get feedback? 

http://forum.kodi.tv/forumdisplay.php?fid=167

**Comment by julianscheel** [comment]
@MartijnKaijser Sounds like a good idea. Can you create a section?

**Comment by ksooo** [comment]
Additionally, it would be a good idea to open the issues section on your github repo so people can start complaining. ;-)

**Comment by ksooo** [comment]
@julianscheel just wanted to confirm that pvr.octonet was built successful on all kodi platforms tonight and thus should be included in all installation sets of latest nightly build (except linux, where we have extra packages for the addons) => http://mirrors.kodi.tv/nightlies/

**Comment by julianscheel** [comment]
@ksoo Thanks for confirming. Is there any further action needed to get it build in the kodi nightly ppa as well?

**Comment by ksooo** [comment]
>  Is there any further action needed to get it build in the kodi nightly ppa as well?

@wsnipex ?

Please try to write my nick name correct, ksoo (with only two "o") is not me. ;-)

**Comment by Jalle19** [comment]
Poor ksoo always getting spammed by us :)

**Pull request merged**